Mobile header responsive style fix

istareatscreens opened the pull request:
Added a popup menu using inline style and absolute positioning to prevent menu overlap on smaller screens

amoutonbrady commented:
@istareatscreens Thank you! I suppose this is still a WIP, I just checked it out, out of curiosity. This is what I'm seeing:

![image](https://user-images.githubusercontent.com/17355226/124511761-2d246880-ddd7-11eb-8a6a-5964ed667f52.png)

On the pure UI side of things, I think I'd like the menu to be like a dropdown menu, right below the button that opens it. I also would like to see labels to make it explicit and compensate the fact that we can't hover over it on mobile.

istareatscreens commented:
@amoutonbrady yes I just wanted to make sure you were fine with what I was doing/have any suggestions before continuing.

I made the changes you suggested, the menu is currently red just to demonstrate its bounds
![image](https://user-images.githubusercontent.com/46049787/124547604-64f0d780-ddfa-11eb-9080-7972ce832f43.png)

-Adjusted some of the icons by adding padding/margin so they are all consistently 24x24px
-I do not know what Icon set you are using for buttons but I suggest a hamburger icon would probably be best. I figure you can add it or provide me the asset tag.
-I used inline style because tailwind doesn't seem have the flexibility I needed to achieve this task. This is my first time using tailwind so if you know a better way to go about this I would love to learn.

amoutonbrady commented:
That's much closer to what I had in mind. I'll give you a proper review
later today. Thank you!

On Tue, Jul 6, 2021, 7:45 AM Stephen Charbonneau ***@***.***>
wrote:

> @amoutonbrady <https://github.com/amoutonbrady> yes I just wanted to make
> sure you were fine with what I was doing/have any suggestions before
> continuing.
>
> I made the changes you suggested, the menu is currently red just to
> demonstrate its bounds
> [image: image]
> <https://user-images.githubusercontent.com/46049787/124547604-64f0d780-ddfa-11eb-9080-7972ce832f43.png>
>
> -Adjusted some of the icons by adding padding/margin so they are all
> consistently 24x24px
> -I do not know what Icon set you are using for buttons but I suggest a
> hamburger icon would probably be best. I figure you can add it or provide
> me the asset tag.
> -I used inline style because tailwind doesn't seem have the flexibility I
> needed to achieve this task. This is my first time using tailwind so if you
> know a better way to go about this I would love to learn.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/solidjs/solid-playground/pull/36#issuecomment-874477698>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AEENDWUUTBSERFHCNATCSDTTWKKA7ANCNFSM47ZBWL5A>
> .
>

amoutonbrady commented:
So, here's a few elements of responses:

- The size if the icon seems pretty fine like this, perhaps a few pixels smaller wouldn't hurt. However, some more padding/margin between the icon and the label would be appreciated. It needs to be the exact same between the boundary and the icon and the icon and the text.
- The icon library is https://heroicons.com/ which is used through a wrapper: https://github.com/amoutonbrady/solid-heroicons. You'll probably find a few burger menus there
- The position of the menu is fine. I'm just a bit worried about overlap with the version dropdown menu.
- I'll make proper code review within the code rather than here

amoutonbrady commented on playground/components/header.tsx:
```diff
@@ -61,88 +71,109 @@ export const Header: Component<{
-                />
+        <div
+          class={`z-10`}
+          style={(showMenu()) ? {
+            position: "absolute",
+            top: "56px", //Set based on header size+margin*2
+            width: "fit-content"
+          } : {}}>
```
This can be written as follow:
```suggestion
         classList={{ 'absolute top-[56px] w-[fit-content]':  showMenu() }}
```

amoutonbrady commented on playground/components/header.tsx:
```diff
@@ -61,88 +71,109 @@ export const Header: Component<{
+              title="Import from JSON"
+            >
+              <input type="file" class="sr-only" onChange={uploadFile} accept=".json" />
+              <Icon path={upload} class="h-6" style={{margin: "0px 0px 0px 0px"}}/>
```
This can be written as follow:
```suggestion
              <Icon path={upload} class="h-6 m-0" />
```

istareatscreens replied:
This doesn't work even with !important enabled, still places weird left margin. I set it back to style={{margin: 0}}. Can't figure out how to do this with tailwind.

istareatscreens replied:
This doesn't work even with !important enabled, still places weird left margin. I set it back to style={{margin: 0}}. Can't figure out how to do this with tailwind.

amoutonbrady commented on playground/components/header.tsx:
```diff
@@ -61,88 +71,109 @@ export const Header: Component<{
+              class="flex flex-row space-x-1 items-center px-3 py-2 focus:outline-none focus:ring-1 rounded text-white opacity-80 hover:opacity-100"
+              title="Export to CodeSandbox"
+            >
+              <svg class="fill-current h-6" style={{ padding: "0px 1.63px 0px 1.63px" }} preserveAspectRatio="xMidYMid" viewBox="0 0 256 296">
```
This can be written as follow, although I'm not entirely sure about the `1.63px`:
```suggestion
              <svg class="fill-current h-6 px-[1.63px] py-0" preserveAspectRatio="xMidYMid" viewBox="0 0 256 296">
```

amoutonbrady commented on playground/components/header.tsx:
```diff
@@ -61,88 +71,109 @@ export const Header: Component<{
-          class="px-3 py-2 focus:outline-none focus:ring-1 rounded text-white opacity-80 hover:opacity-100"
-          title="Export to JSON"
+          id="menu-btn"
+          class="z-40 px-3 py-2 focus:outline-none focus:ring-1 rounded text-white opacity-80 hover:opacity-100 visible relative md:hidden"
+          style={{margin:"0 10px 0 0"}} //adjust button to the right to better algin menu
```
This can be written as follow, although I generally prefer only margin on the left and top:
```suggestion
          class="z-40 px-3 py-2 focus:outline-none focus:ring-1 rounded text-white opacity-80 hover:opacity-100 visible relative md:hidden m-0 mr-[10px]"
```

istareatscreens commented:
Tried to follow your guidelines. I made changes to attempt to make the menu similar to the version drop down. 

I spaced the text farther from the icons.

![image](https://user-images.githubusercontent.com/46049787/124684301-cc5f6380-de9c-11eb-859c-45228c5c2579.png)

![image](https://user-images.githubusercontent.com/46049787/124684331-d84b2580-de9c-11eb-847f-da7029b040c0.png)

I spent a little time trying to get hover and focus to work in tailwind but no luck. 

Closing the menu by clicking outside of it seems difficult to implement due to Monaco. As a solution I just switch the hamburger icon to a x icon.

amoutonbrady commented on playground/components/header.tsx:
```diff
@@ -62,90 +72,146 @@ export const Header: Component<{
+          class="z-10"
+        >
+          <div
+            class="shadow-md md:items-center md:space-x-2 md:flex"
```
We are missing a `shadow-none` on desktop here
```suggestion
            class="shadow-md md:shadow-none md:items-center md:space-x-2 md:flex"
```

istareatscreens replied:
Moved shadow to classList and cleaned up this in general

amoutonbrady commented on playground/components/header.tsx:
```diff
@@ -62,90 +72,146 @@ export const Header: Component<{
-          >
-            <svg class="h-6" fill="none" viewBox="0 0 24 24" stroke="currentColor">
+          {showMenu() ? (
+            <svg
+              xmlns="http://www.w3.org/2000/svg"
+              class="h-6 w-6"
+              fill="none"
+              viewBox="0 0 24 24"
+              stroke="currentColor"
+            >
```
We can replace the whole SVG by the icon from `@amoutonbrady/solid-heroicons`:
```suggestion
            <Icon path={xCircle} class="h-6 w-6" />
```

istareatscreens replied:
Done

amoutonbrady commented on playground/components/header.tsx:
```diff
@@ -62,90 +72,146 @@ export const Header: Component<{
-          <span class="sr-only">{copy() ? 'Copied to clipboard' : 'Share'}</span>
-          <Icon class="h-6" path={copy() ? link : share} />
+          ) : (
+            <svg
+              xmlns="http://www.w3.org/2000/svg"
+              class="h-6 w-6"
+              fill="none"
+              viewBox="0 0 24 24"
+              stroke="currentColor"
+            >
+              <path
+                stroke-linecap="round"
+                stroke-linejoin="round"
+                stroke-width={2}
+                d="M4 6h16M4 10h16M4 14h16M4 18h16"
+              />
+            </svg>
```
Same as previous comment:
```suggestion
            <Icon path={menu} class="h-6 w-6" />
```

istareatscreens replied:
Done

amoutonbrady commented on playground/components/header.tsx:
```diff
@@ -62,90 +72,146 @@ export const Header: Component<{
-            }
-          >
-            <svg class="h-6" fill="none" viewBox="0 0 24 24" stroke="currentColor">
+          {showMenu() ? (
```
We can replace this ternary by the `<Show />` component:
```suggestion
          <Show when={showMenu()} fallback={<Icon path={menu}  class="h-6 w-6" />}> 
              <Icon path={xCircle}  class="h-6 w-6" />
          </Show>
```

istareatscreens replied:
Done

amoutonbrady commented on tailwind.config.js:
```diff
@@ -10,6 +10,22 @@ module.exports = {
   ],
   theme: {
     extend: {
+      screens: {
+        'sm': '640px',
+        // => @media (min-width: 640px) { ... }
+  
+        'md': '768px',
+        // => @media (min-width: 768px) { ... }
+  
+        'lg': '1024px',
+        // => @media (min-width: 1024px) { ... }
+  
+        'xl': '1280px',
+        // => @media (min-width: 1280px) { ... }
+  
+        '2xl': '1536px',
+        // => @media (min-width: 1536px) { ... }
+      },
```
We do not need those. They are in by default.
```suggestion
```

istareatscreens replied:
Committed

amoutonbrady reviewed (commented):
That's getting a whole lot closer to what I pictured. That being said, I've noticed a couple of optimizations possible. I do also notice a weird change of header height when clicking the icon back and forth.

You are definitely going in the right direction. I didn't see those hover styles you were mentioning though.

istareatscreens commented:
I styled the buttons in the menu and fixed the header size increase bug, by resizing the xClose button to 22x22 (to make up for the white border added).

I changed what Icons I could to svg, including moon and sun

Removed all ternary operators I could with Show as directed

amoutonbrady commented:
This is perfect. Thank you for you work and time!
